lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -517,31 +525,52 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
                             liftIO $ traceWith tr $ MsgErrProduction e
                         Right () ->
                             pure ()
-
-                -- A single block can contain multiple certificates relating to the
-                -- same pool.
-                --
-                -- The /order/ in which certificates appear is /significant/:
-                -- certificates that appear later in a block /generally/ take
-                -- precedence over certificates that appear earlier on.
-                --
-                -- We record /all/ certificates within the database, together with
-                -- the order in which they appeared.
-                --
-                -- Precedence is determined by the 'readPoolLifeCycleStatus'
-                -- function.
-                --
-                let publicationTimes =
-                        CertificatePublicationTime slot <$> [minBound ..]
-                forM_ (publicationTimes `zip` certificates) $ \case
-                    (publicationTime, Registration cert) -> do
-                        liftIO $ traceWith tr $ MsgStakePoolRegistration cert
-                        putPoolRegistration publicationTime cert
-                    (publicationTime, Retirement cert) -> do
-                        liftIO $ traceWith tr $ MsgStakePoolRetirement cert
-                        putPoolRetirement publicationTime cert
+                unless (null certificates) $ do
+                    -- Before adding new pool certificates to the database, we
+                    -- first attempt to garbage collect pools that have already
+                    -- retired. By only performing garbage collection when we
+                    -- actually have new certificates to add, we avoid paying
+                    -- the cost of garbage collection in every single slot.
+                    garbageCollectPools slot
+                    putPoolCertificates slot certificates
         pure Continue
 
+    -- Perform garbage collection for pools that have retired.
+    --
+    -- To avoid any issues with rollback, we err on the side of caution and
+    -- only garbage collect pools that retired /at least/ two epochs before
+    -- the current epoch.
```

Comment:
What's the rationale for "at least 2". Why not, "at least 3" for example?

Reply:
> What's the rationale for "at least 2". Why not, "at least 3" for example?

If we've **only just** crossed the boundary from epoch **_e_** into epoch **_e_** **+ 1**, this means that blocks from the previous epoch are still within the rollback window. If a rollback does happen, and if the chain we've just been following has been replaced with another chain in which the retirement certificate for some pool **p** is revoked within epoch **_e_** (with a registration certificate that supersedes it), then the retirement for pool **p** would be cancelled.

But assuming the rollback window is guaranteed to be less than one epoch in length, retirements that occurred two epochs ago can no longer be affected by rollback.

Reply:
Would it be possible to extend the comment with this rationale :) ?

Reply:
> Would it be possible to extend the comment with this rationale :) ?

Certainly!

Fixed in 395a64020d9ed0c92d53d0cd4de2d9a38e29a0f9

Reply:
👍